### PR TITLE
unify `shuffle` and `randperm`

### PR DESCRIPTION
**Pull request by ctarn**
unify the algorithms of `shuffle` and `randperm` without performance loss, so that:
- `shuffle(rng, 1:n) == randperm(rng, n)`
- `shuffle(rng, eachindex(1:n)) == shuffle(rng, 1:n)`

see also:
- https://github.com/JuliaLang/julia/issues/50304
- https://github.com/JuliaLang/julia/pull/45732

correctness:
```julia
ns = [1, 10, 100, 1000, 10000]

for n in ns
    @show shuffle!(MersenneTwister(1), Vector(1:n) .* 7) ./ 7 == randperm(MersenneTwister(1), n)
end

"""
output:
shuffle!(MersenneTwister(1), Vector(1:n) .* 7) ./ 7 == randperm(MersenneTwister(1), n) = true
shuffle!(MersenneTwister(1), Vector(1:n) .* 7) ./ 7 == randperm(MersenneTwister(1), n) = true
shuffle!(MersenneTwister(1), Vector(1:n) .* 7) ./ 7 == randperm(MersenneTwister(1), n) = true
shuffle!(MersenneTwister(1), Vector(1:n) .* 7) ./ 7 == randperm(MersenneTwister(1), n) = true
shuffle!(MersenneTwister(1), Vector(1:n) .* 7) ./ 7 == randperm(MersenneTwister(1), n) = true
"""
```

performance:
```julia
ns = [1, 10, 100, 1000, 10000]

t_new = [@belapsed(shuffle!(MersenneTwister(1), Vector(1:$n))) for n in ns]
t_old = [@belapsed(Random.shuffle!(MersenneTwister(1), Vector(1:$n))) for n in ns]
println("small item:\n#item\tnew\told")
display(hcat(ns, t_new, t_old))

"""
output:
small item:
#item   new     old
5×3 Matrix{Float64}:
     1.0  9.75e-6    9.748e-6
    10.0  9.938e-6   1.02e-5
   100.0  1.0523e-5  1.0548e-5
  1000.0  1.3765e-5  1.3549e-5
 10000.0  7.2876e-5  7.1795e-5
"""

vs = [[NTuple{1000, Float64}(rand(1000)) for _ in 1:n] for n in ns]
t_new = [@belapsed(shuffle!(MersenneTwister(1), $v)) for v in copy.(vs)]
t_old = [@belapsed(Random.shuffle!(MersenneTwister(1), $v)) for v in copy.(vs)]
println("large item:\n#item\tnew\told")
display(hcat(ns, t_new, t_old))

"""
output:
large item:
#item   new     old
5×3 Matrix{Float64}:
     1.0  9.811e-6     9.733e-6
    10.0  1.2482e-5    1.2365e-5
   100.0  4.4791e-5    4.3842e-5
  1000.0  0.000401874  0.000391641
 10000.0  0.00989083   0.0101406
"""
```

### CODE REVIEW

**Comment by stevengj** [comment]
It makes sense not to me not to have `shuffle` and `randperm` use gratuitously different algorithms.   Would be good for someone who implemented these to take a look, however.

**Comment by ctarn** [comment]
The difference can be traced back to the initial implementation of 13 years ago by @ViralBShah:

https://github.com/JuliaLang/julia/blob/18b72d1c565dca19696ef5ad0d07060cf7b7620a/combinatorics.j#L151-L169

@rfourquet made an improvement in #14772 while preserving the original style.

I don't see any reason why they are different, and since `shuffle` on `Base.OneTo(n)` is currently specialized as `randperm(n)`, it is confusing that `shuffle(collect(Base.OneTo(n))) != shuffle(Base.OneTo(n))` while `collect(Base.OneTo(n)) == Base.OneTo(n)`.

https://github.com/JuliaLang/julia/blob/e4600c5f5b172963e5394dad831204e69c74d949/stdlib/Random/src/misc.jl#L267

```julia
julia> n = 4

julia> shuffle(MersenneTwister(1), collect(Base.OneTo(n)))
4-element Vector{Int64}:
 1
 2
 4
 3

julia> shuffle(MersenneTwister(1), Base.OneTo(n))
4-element Vector{Int64}:
 1
 3
 4
 2

julia> collect(Base.OneTo(n)) == Base.OneTo(n)
true

julia> shuffle(MersenneTwister(1), collect(Base.OneTo(n))) == shuffle(MersenneTwister(1), Base.OneTo(n))
false
```

**Comment by Seelengrab** [comment]
> * shuffle(rng, 1:n) == randperm(rng, n)
> * shuffle(rng, eachindex(1:n)) == shuffle(rng, 1:n)

Why is it advanteguous to guarantee that these properties hold? It's all randomness either way - I don't think guaranteeing a particular result is desirable.

**Comment by oscardssmith** [comment]
It's not a major advantage, but if it's easy to do it seems worthwhile

**Comment by stevengj** [comment]
For one thing, it makes the source code less readable/maintainable if `shuffle` and `randperm` use gratuitously different algorithms.

**Review by rfourquet — APPROVED** [review]
LGTM

File: stdlib/Random/src/misc.jl [hunk 1]
```diff
@@ -326,7 +326,7 @@ function randperm!(r::AbstractRNG, a::Array{<:Integer})
             a[i] = a[j]
         end
         a[j] = i
-        i == 1+mask && (mask = 2mask + 1)
+        i == 1 + mask && (mask = 2 * mask + 1)
```
Comment (stevengj) [inline stdlib/Random/src/misc.jl:329]: Might want to do a similar reformatting in `randcycle!`

**Comment by Seelengrab** [comment]
If we want to guarantee these properties, this PR also should check that that they hold in a test.

File: stdlib/Random/src/misc.jl (hunk 1 above)
Comment (ctarn) [reply stdlib/Random/src/misc.jl:331]: Of course. I also added notes to remind people to keep the consistency.

**Comment by ctarn** [comment]
> If we want to guarantee these properties, this PR also should check that that they hold in a test.

It can prevent some further improvements, although this is unlikely to happen. I am happy to add some tests, but don't insist on it.

These properties can also be documented, but don't have to be, I think.

**Pull request merged**